File: 189. Rotate Array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        //more runtime
        if (nums.size() < 2)
            return;
        
        k %= nums.size();
        reverse(nums, 0, nums.size() - 1);
        reverse(nums, 0, k -1);
        reverse(nums, k, nums.size() - 1);
    }
    inline void reverse(vector<int>& nums, int begin, int end) {
        while (begin < end) {
            swap(nums.at(begin), nums.at(end));
            begin++;
            end--;
        }
    }
    inline void swap(int& a, int& b) {
        int tmp = a;
        a = b;
        b = tmp;
    }
};

File: java/189. Rotate Array.java
// 189. Rotate Array
// TC: O(N) where N = nums length
// SC: O(1)

class Solution {
    public void rotate(int[] nums, int k) {
        int pending_element = nums.length;
        int finished_element = 0;
        int first = 0;
        int second = 0;
        while (pending_element > 0 && (k %= pending_element) >= 1) {
            for (int i = 0; i < k; i++) {
                //swap
                first = i + finished_element;
                second = pending_element + finished_element - k + i;
                nums[first] ^= nums[second];
                nums[second] ^= nums[first];
                nums[first] ^= nums[second];
            }
            finished_element += k;
            pending_element -= k;
        }
    }
}